src/library/export/trackexportworker.cpp

```diff
@@ -71,9 +70,31 @@ QMap<QString, TrackFile> createCopylist(const TrackPointerList& tracks) {
 
 }  // namespace
 
+QList<QString> TrackExportWorker::getMissingTracks() {
+    QList<QString> missingFiles;
+    for (const auto& it : m_tracks) {
+        if (it->getCanonicalLocation().isEmpty()) {
+            missingFiles << it->getFileInfo().fileName();
+        }
+    }
+    missingFiles.removeDuplicates();
+    return missingFiles;
+}
+
 void TrackExportWorker::run() {
     int i = 0;
     QMap<QString, TrackFile> copy_list = createCopylist(m_tracks);
+
+    /**
+     * TODO: if the copylist is empty, then end the process
+     */
+
+    if (copy_list.empty()) {
+        emit canceled();
+        m_errorMessage = tr("There were no files in the specified file locations to export!");
```

Comment:
The error message will not be seen by receivers of the canceled() signal

Reply:
Do they need to see it? I wanted to implement some sort of pop up message for the user to see that there are no files to export, and then close the thread. emit canceled() seemed like an easy way to get it done. From a UI perspective it did the job for me. Do you have a suggestion for another way to implement the functionality?

Reply:
Only emit signals after updating the internal state has been finished. Just switch the ordering.